src/Psl/Xml/detect_errors.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Xml;
+
+use LibXMLError;
+use Psl\Xml\Exception\XmlErrorsException;
+
+use function libxml_clear_errors;
+use function libxml_get_errors;
+use function libxml_use_internal_errors;
+
+/**
+ * You can wrap this detector around a function that interacts with xml, dom, simplexml, ...
+ * It wil trigger a `XmlErrorsException` if an error was encountered.
+ * No XML warnings will be displayed in the meantime.
+ * The result of your function will be returned as result.
+ *
+ * @template Tr
+ *
+ * @psalm-param callable(): Tr $run
+ * @psalm-return Tr
+ *
+ * @throws XmlErrorsException
+ */
+function detect_errors(callable $run)
+{
+    $previousErrorReporting = libxml_use_internal_errors(true);
+    libxml_clear_errors();
+    $result = $run();
+
+    /** @var list<LibXMLError> $errors */
+    $errors = libxml_get_errors();
+    libxml_clear_errors();
+    libxml_use_internal_errors($previousErrorReporting);
+
+    if ($errors) {
```

Comment:
Not sure if this is the best way. For example : whilst validating XSD schema's, it is possible to receive warnings.
Should those be 'blocking'? It's not that it results in a completely unusable XML format.

How would we be able to deal with this? I was thinking about either

- Adding a parameter to set an error level
- return a tupple so that you can have both a result and/or an exception - similar to [go's error handling](https://blog.golang.org/error-handling-and-go)

WDYT?

Reply:
we should be as strict as possible here, i think throwing an exception is the right thing to do, however, throwing an exception that is compsed of multiple errors seems wrong, we should throw only for the first error.

`$error = Arr\first($errors)` and we throw that.

we can add a error level parameter, i don't like the API but i think it would be okay since this should be in the internal namespace, where we can have strictness configured in the public api.

e.g:

$parser = new Xml\Parser(strict: true); // true => all, false => ignore warnings.

Reply:
I uploaded an alternative solution which is IMO the way to go:

The `$run` method can either succeed or fail, so is a result.
Next to that result : you get a list of errors, which is the context on why something failed or like warnings when the run method succeeded.

An Example can be seen in phpunit : if you add invalid xml tag, phpunit will display e.g.

```
  Warning - The configuration file did not pass validation!
  The following problems have been detected:

  Line 29:
  - Element 'someinvaliddtag': This element is not expected.

  Test results may not be as expected.
```

However, it still works as expected.


Only returning the first error is not OK, since E.g. parsing an xsd may result in multiple issues. One might be more blocking than another one.


By doing it the new way, the implementing function is in full charge on how to deal with exceptions.
For example:

- load xml 
  - The load function will throw an exception : cannot load, the errors collection will explain why.
  - The load function can return a document, but with some notices to display (like phpunit)
- validate xml
  - The implementation can determine which level is blocking, the detect function just returns everything.
  - The validation bool result could be changed based on errors - if it makes sense